### PR TITLE
Introduce hsec-sync (#166)

### PR DESCRIPTION
**Pull request by blackheaven**
I have create a new library/executable to synchronize a local security advisories.

Maybe I should also add a `status` operation, WDYT?

See #166 

Maybe be use in #148 and by flora (/cc @tchoutri).

### CODE REVIEW

File: code/hsec-sync/hsec-sync.cabal [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+license:         BSD-3-Clause
+author:          Gautier DI FOLCO
+maintainer:      gautier.difolco@gmail.com
+
+-- A copyright notice.
+-- copyright:
+category:        Data
+extra-doc-files: CHANGELOG.md
+tested-with:
+  GHC ==8.10.7 || ==9.0.2 || ==9.2.8 || ==9.4.8 || ==9.6.3 || ==9.8.1
```
Comment (Kleidukos) [inline code/hsec-sync/hsec-sync.cabal:24]: Not sure if it's wise to support so many GHC versions for a CLI tool. I would advise that you stay on the highest supported GHC at all times. Most people will anyway take your release binaries, especially in CI.

File: code/hsec-sync/src/Security/Advisories/Sync/Git.hs [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+          then Left $ GitError cmd $ GitProcessError status stdout stderr
+          else Right GitRepositoryCreated
+
+updateGitRepository :: Repository -> IO (Either GitError ())
+updateGitRepository repo =
+  D.withCurrentDirectory (repositoryRoot repo) $ do
+    _ <- runGit ["remote", "add", "origin", repositoryUrl repo] -- can fail if it exists
+    (setUrlCmd, setUrlStatus, setUrlStdout, setUrlStderr) <-
+      runGit ["remote", "set-url", "origin", repositoryUrl repo]
+    if setUrlStatus /= ExitSuccess
```
Comment (Kleidukos) [inline code/hsec-sync/src/Security/Advisories/Sync/Git.hs:122]: I believe that some, if not most, of these early returns can be replaced with `when`. (also you should be using `pure`, for reasons of forward-compatibility.)

**Comment by MangoIV** [comment]
for me this is okay as long as I have a directory or git repository url to point it at.

File: code/hsec-sync/hsec-sync.cabal (hunk 1 above)
Comment (blackheaven) [reply code/hsec-sync/hsec-sync.cabal:24]: Actually, CLI is more a bonus, I had flora/hackage-server in mind when I have made it.

File: code/hsec-sync/src/Security/Advisories/Sync/Git.hs (hunk 2 above)
Comment (blackheaven) [reply code/hsec-sync/src/Security/Advisories/Sync/Git.hs:122]: IIUC `when` does not shortcut/early return in `IO`.

Comment (Kleidukos) [reply code/hsec-sync/src/Security/Advisories/Sync/Git.hs:122]: So, for context: they could be replaced by when guards and have the appropriate early return effect if there was a mechanism of exception handling that existed. Very much something I expect from my usual web application business in Haskell so much that I didn't realise I had come to expect the same structure everywhere else without explanation. (I will not move an inch on this matter of pure vs. return but you know I love you dearly nonetheless)

Comment (MangoIV) [reply code/hsec-sync/src/Security/Advisories/Sync/Git.hs:122]: You could just use ExceptT

Comment (MangoIV) [reply code/hsec-sync/src/Security/Advisories/Sync/Git.hs:122]:
```haskell
bla :: IO ()
bla =  print =<< runExceptT do
  when True do 
    throwError "oh no" 
  liftIO $ print "hello"
```

I agree the nesting is not beautiful...

**Pull request merged**